tests/gtest/avifallocationtest.cc

```diff
@@ -0,0 +1,187 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include <limits>
+#include <vector>
+
+#include "avif/avif.h"
+#include "aviftest_helpers.h"
+#include "gtest/gtest.h"
+
+using ::testing::Combine;
+using ::testing::Values;
+using ::testing::ValuesIn;
+
+namespace libavif {
+namespace {
+
+void TestAllocation(uint32_t width, uint32_t height, uint32_t depth,
+                    avifResult expected_result) {
+  // The format of the image and which planes are allocated should not matter.
+  // Test all combinations.
+  for (avifPixelFormat format :
+       {AVIF_PIXEL_FORMAT_NONE, AVIF_PIXEL_FORMAT_YUV444,
+        AVIF_PIXEL_FORMAT_YUV422, AVIF_PIXEL_FORMAT_YUV420,
+        AVIF_PIXEL_FORMAT_YUV400}) {
+    for (avifPlanesFlag planes :
+         {AVIF_PLANES_YUV, AVIF_PLANES_A, AVIF_PLANES_ALL}) {
+      testutil::AvifImagePtr image(avifImageCreateEmpty(), avifImageDestroy);
+      ASSERT_NE(image, nullptr);
+      image->width = width;
+      image->height = height;
+      image->depth = depth;
+      image->yuvFormat = format;
+      EXPECT_EQ(avifImageAllocatePlanes(image.get(), planes), expected_result);
+
+      // Make sure the actual plane pointers are consistent with the settings.
+      if (expected_result == AVIF_RESULT_OK &&
+          format != AVIF_PIXEL_FORMAT_NONE && (planes & AVIF_PLANES_YUV)) {
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_Y], nullptr);
+      } else {
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_Y], nullptr);
+      }
+      if (expected_result == AVIF_RESULT_OK &&
+          format != AVIF_PIXEL_FORMAT_NONE &&
+          format != AVIF_PIXEL_FORMAT_YUV400 && (planes & AVIF_PLANES_YUV)) {
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_U], nullptr);
+        EXPECT_NE(image->yuvPlanes[AVIF_CHAN_V], nullptr);
+      } else {
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_U], nullptr);
+        EXPECT_EQ(image->yuvPlanes[AVIF_CHAN_V], nullptr);
+      }
+      if (expected_result == AVIF_RESULT_OK && (planes & AVIF_PLANES_A)) {
+        EXPECT_NE(image->alphaPlane, nullptr);
+      } else {
+        EXPECT_EQ(image->alphaPlane, nullptr);
+      }
+    }
+  }
+}
+
+TEST(AllocationTest, MinimumValidDimensions) {
+  TestAllocation(1, 1, 8, AVIF_RESULT_OK);
+}
+
+TEST(AllocationTest, MaximumValidDimensions) {
+  // On 32-bit builds, malloc() will fail with fairly low sizes.
+  // Adapt the tests to take that into account.
+  constexpr bool kIsPlatform64b = sizeof(void*) > 4;
+  constexpr uint32_t kMaxAllocatableDimension =
+      kIsPlatform64b ? std::numeric_limits<typeof(avifImage::width)>::max()
```

Comment:
Yannis: When I tried to compile libavif with MSVC, I found that `typeof` is a GCC extension. It is not supported by MSVC.

Reply:
Fixed in https://github.com/AOMediaCodec/libavif/pull/1014